hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/hudi/SparkAdapter.scala

```diff
@@ -250,4 +260,52 @@ trait SparkAdapter extends Serializable {
    * @return
    */
   def stopSparkContext(jssc: JavaSparkContext, exitCode: Int): Unit
+
+  def getSchema(conn: Connection,
+                resultSet: ResultSet,
+                dialect: JdbcDialect,
+                alwaysNullable: Boolean = false,
+                isTimestampNTZ: Boolean = false): StructType
+
+  /**
+   * [SPARK-46832] Using UTF8String compareTo directly would throw UnsupportedOperationException since Spark 4.0
+   */
+  def compareUTF8String(a: UTF8String, b: UTF8String): Int = a.compareTo(b)
+
+  /**
+   * [SPARK-46832] Using UTF8String compareTo directly would throw UnsupportedOperationException since Spark 4.0
+   * FlatLists is a static class and we cannot override any methods within to change the logic for comparison
+   * So we have to create [[Spark4FlatLists]] for Spark 4.0+
+   */
+  def createComparableList(t: Array[AnyRef]): FlatLists.ComparableList[Comparable[HoodieRecord[_]]] = FlatLists.ofComparableArray(t)
```

Comment:
For places that need to compare `UTF8String`, could we introduce `HoodieUTF8String` implemented by `Spark3HoodieUTF8String` and `Spark4HoodieUTF8String` and override `compareTo` by calling the right method in different Spark versions, so that we can avoid having many places using spark adapter?

For example, here it would receive an array that can contain `HoodieUTF8String` which works with `compareTo` so there is no need to change `FlatLists` and add `Spark4FlatLists` (which shares the same code mostly).